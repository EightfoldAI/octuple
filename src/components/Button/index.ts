export * from './Button.types';
export * from './BaseButton';
export * from './DefaultButton/DefaultButton';
export * from './NeutralButton/NeutralButton';
export * from './PrimaryButton/PrimaryButton';
export * from './SecondaryButton/SecondaryButton';
export * from './SplitButton/SplitButton';
export * from './TwoStateButton/TwoStateButton';
