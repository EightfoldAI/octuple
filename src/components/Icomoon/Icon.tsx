import React, { FC, useContext } from 'react';
import { mergeClasses } from '../../shared/utilities';
import { IcomoonProps, IcomoonSize } from './';
import IcomoonReact from 'icomoon-react';

import styles from './icon.module.scss';

const IconSetContext = React.createContext<object | null>(null);

export const IcomoonProvider = IconSetContext.Provider;

export const Icomoon: FC<IcomoonProps> = ({
    ariaHidden = false,
    classNames,
    color,
    iconName,
    id,
    role = 'presentation',
    size = IcomoonSize.Medium,
    'data-test-id': dataTestId,
}) => {
    const iconClassNames: string = mergeClasses([
        classNames,
        styles.iconWrapper,
    ]);

    return (
        <span
            data-test-id={dataTestId}
            aria-hidden={ariaHidden}
            className={iconClassNames}
            id={id}
            role={role}
        >
            <IconSetContext.Consumer>
                {(iconSet) => (
                    <IcomoonReact
                        iconSet={iconSet}
                        icon={iconName}
                        size={size}
                        color={color}
                    />
                )}
            </IconSetContext.Consumer>
        </span>
    );
};
