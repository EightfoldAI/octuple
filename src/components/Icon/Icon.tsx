'use client';

import React, { FC } from 'react';
import IcomoonReact from 'icomoon-react';

import { mergeClasses } from '../../shared/utilities';
import { Icon as MdiIcon } from '@mdi/react';
import { IconProps, IconSize } from './index';

import styles from './icon.module.scss';
import { useConfig } from '../ConfigProvider';

export const Icon: FC<IconProps> = ({
  ariaHidden = false,
  classNames,
  color,
  description,
  horizontal,
  id,
  path,
  role = 'presentation',
  rotate,
  size = IconSize.Medium,
  spin,
  style,
  title,
  vertical,
  'data-test-id': dataTestId,
  icomoonIconName,
}) => {
  const { icomoonIconSet } = useConfig();

  const iconClassNames: string = mergeClasses([classNames, styles.iconWrapper]);

  if (!icomoonIconName && !path) {
    return null;
  }

  const iconComponent = icomoonIconName ? (
    <IcomoonReact
      iconSet={icomoonIconSet}
      size={size}
      color={color}
      icon={icomoonIconName}
      role="presentation"
    />
  ) : (
    <MdiIcon
      color={color}
      description={description}
      horizontal={horizontal}
      path={path}
      rotate={rotate}
      size={size}
      title={title}
      vertical={vertical}
      spin={spin}
    />
  );

  return (
    <span
      data-test-id={dataTestId}
      aria-hidden={ariaHidden}
      className={iconClassNames}
      id={id}
      role={role}
      style={style ? style : null}
    >
      {iconComponent}
    </span>
  );
};

Icon.displayName = 'Icon';
