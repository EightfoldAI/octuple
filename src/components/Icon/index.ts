export * from './Icon';
export * from './Icon.types';
export * from './mdi';
