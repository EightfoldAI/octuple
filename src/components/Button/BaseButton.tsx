import React, { FC, Ref } from 'react';
import {
    ButtonIconAlign,
    ButtonShape,
    ButtonSize,
    ButtonTextAlign,
    ButtonTheme,
    ButtonWidth,
    InternalButtonProps,
    SplitButton,
} from './';
import { Icon, IconSize } from '../Icon';
import { Badge } from '../Badge';
import { Breakpoints, useMatchMedia } from '../../hooks/useMatchMedia';
import { mergeClasses } from '../../shared/utilities';

import styles from './button.module.scss';

export const BaseButton: FC<InternalButtonProps> = React.forwardRef(
    (
        {
            alignIcon = ButtonIconAlign.Left,
            alignText = ButtonTextAlign.Center,
            allowDisabledFocus = false,
            ariaLabel,
            buttonWidth = ButtonWidth.fitContent,
            checked = false,
            classNames,
            counter,
            disabled = false,
            disruptive = false,
            dropShadow = false,
            htmlType,
            iconProps,
            id,
            onClick,
            onContextMenu,
            shape = ButtonShape.Pill,
            size = ButtonSize.Flex,
            split,
            splitButtonChecked = false,
            splitButtonProps,
            style,
            text,
            theme,
            toggle,
            type,
            ...rest
        },
        ref: Ref<HTMLButtonElement>
    ) => {
        const largeScreenActive: boolean = useMatchMedia(Breakpoints.Large);
        const mediumScreenActive: boolean = useMatchMedia(Breakpoints.Medium);
        const smallScreenActive: boolean = useMatchMedia(Breakpoints.Small);
        const xSmallScreenActive: boolean = useMatchMedia(Breakpoints.XSmall);

        const counterExists: boolean = !!counter;
        const iconExists: boolean = !!iconProps;
        const textExists: boolean = !!text;

        const buttonBaseSharedClassNames: string = mergeClasses([
            classNames,
            {
                [styles.buttonSize3]:
                    size === ButtonSize.Flex && largeScreenActive,
            },
            {
                [styles.buttonSize2]:
                    size === ButtonSize.Flex && mediumScreenActive,
            },
            {
                [styles.buttonSize2]:
                    size === ButtonSize.Flex && smallScreenActive,
            },
            {
                [styles.buttonSize1]:
                    size === ButtonSize.Flex && xSmallScreenActive,
            },
            { [styles.buttonSize1]: size === ButtonSize.Large },
            { [styles.buttonSize2]: size === ButtonSize.Medium },
            { [styles.buttonSize3]: size === ButtonSize.Small },
            { [styles.pillShape]: shape === ButtonShape.Pill },
            { [styles.dropShadow]: dropShadow },
            { [styles.dark]: theme === ButtonTheme.dark },
        ]);

        const buttonBaseClassNames: string = mergeClasses([
            buttonBaseSharedClassNames,
            { [styles.buttonStretch]: buttonWidth === ButtonWidth.fill },
            { [styles.splitLeft]: split },
            { [styles.left]: alignText === ButtonTextAlign.Left },
            { [styles.right]: alignText === ButtonTextAlign.Right },
            {
                [styles.iconLeft]:
                    iconExists && alignIcon === ButtonIconAlign.Left,
            },
            {
                [styles.iconRight]:
                    iconExists && alignIcon === ButtonIconAlign.Right,
            },
            { [styles.disabled]: allowDisabledFocus || disabled },
        ]);

        const buttonTextClassNames: string = mergeClasses([
            {
                [styles.buttonText3]:
                    size === ButtonSize.Flex && largeScreenActive,
            },
            {
                [styles.buttonText2]:
                    size === ButtonSize.Flex && mediumScreenActive,
            },
            {
                [styles.buttonText2]:
                    size === ButtonSize.Flex && smallScreenActive,
            },
            {
                [styles.buttonText1]:
                    size === ButtonSize.Flex && xSmallScreenActive,
            },
            { [styles.buttonText1]: size === ButtonSize.Large },
            { [styles.buttonText2]: size === ButtonSize.Medium },
            { [styles.buttonText3]: size === ButtonSize.Small },
        ]);

        const badgeClassNames: string = mergeClasses([
            styles.counter,
            buttonTextClassNames,
        ]);

        const getButtonIconSize = (): IconSize => {
            let iconSize: IconSize;
            if (size === ButtonSize.Flex && largeScreenActive) {
                iconSize = IconSize.Small;
            } else if (
                size === ButtonSize.Flex &&
                (mediumScreenActive || smallScreenActive)
            ) {
                iconSize = IconSize.Medium;
            } else if (size === ButtonSize.Flex && xSmallScreenActive) {
                iconSize = IconSize.Large;
            } else if (size === ButtonSize.Large) {
                iconSize = IconSize.Large;
            } else if (size === ButtonSize.Medium) {
                iconSize = IconSize.Medium;
            } else if (size === ButtonSize.Small) {
                iconSize = IconSize.Small;
            }
            return iconSize;
        };

        const getButtonIcon = (): JSX.Element => (
            <Icon
                {...iconProps}
                classNames={mergeClasses([styles.icon, iconProps.classNames])}
                size={getButtonIconSize()}
            />
        );

        const getButtonContent = (
            buttonTextClassNames: string,
            text: string
        ): JSX.Element => (
            <span className={buttonTextClassNames}>
                {text ? text : 'Button'}
                {counterExists && (
                    <Badge classNames={badgeClassNames}>{counter}</Badge>
                )}
            </span>
        );

        return (
            <>
                <button
                    ref={ref}
                    {...rest}
                    aria-checked={toggle ? !!checked : undefined}
                    aria-disabled={allowDisabledFocus}
                    aria-label={ariaLabel}
                    aria-pressed={toggle ? !!checked : undefined}
                    defaultChecked={checked}
                    disabled={disabled}
                    className={buttonBaseClassNames}
                    id={id}
                    onClick={!allowDisabledFocus ? onClick : null}
                    style={style}
                    type={htmlType}
                >
                    {iconExists && !textExists && getButtonIcon()}
                    {counterExists && !textExists && (
                        <Badge classNames={badgeClassNames}>{counter}</Badge>
                    )}
                    {iconExists && textExists && (
                        <span>
                            {getButtonIcon()}
                            {getButtonContent(buttonTextClassNames, text)}
                        </span>
                    )}
                    {!iconExists &&
                        getButtonContent(buttonTextClassNames, text)}
                </button>
                {split && (
                    <SplitButton
                        {...splitButtonProps}
                        classNames={
                            buttonBaseSharedClassNames +
                            ' ' +
                            splitButtonProps?.classNames
                        }
                        checked={splitButtonChecked}
                        disruptive={disruptive}
                        dropShadow={dropShadow}
                        onClick={
                            !splitButtonProps?.allowDisabledFocus
                                ? onContextMenu
                                : null
                        }
                        shape={shape}
                        size={size}
                        split={split}
                        theme={theme}
                        type={type}
                    />
                )}
            </>
        );
    }
);
