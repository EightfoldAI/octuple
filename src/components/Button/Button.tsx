import React, { FC, Ref, useContext, useEffect, useRef, useState } from 'react';
import DisabledContext, { Disabled } from '../ConfigProvider/DisabledContext';
import { SizeContext, Size } from '../ConfigProvider';
import {
  ButtonIconAlign,
  ButtonProps,
  ButtonShape,
  ButtonSize,
  ButtonTextAlign,
  ButtonVariant,
  ButtonWidth,
  SplitButton,
} from '.';
import { Icon, IconSize } from '../Icon';
import { Badge } from '../Badge';
import { InnerNudge, NudgeAnimation, NudgeProps } from './Nudge';
import { Breakpoints, useMatchMedia } from '../../hooks/useMatchMedia';
import { mergeClasses } from '../../shared/utilities';
import { Loader, LoaderSize } from '../Loader';
import { useCanvasDirection } from '../../hooks/useCanvasDirection';
import { useMergedRefs } from '../../hooks/useMergedRefs';
import { useNudge } from './Nudge/Hooks/useNudge';

import styles from './button.module.scss';

export const Button: FC<ButtonProps> = React.forwardRef(
  (props: ButtonProps, ref: Ref<HTMLButtonElement>) => {
    const {
      alignIcon = ButtonIconAlign.Left,
      alignText = ButtonTextAlign.Center,
      allowDisabledFocus = false,
      ariaLabel,
      buttonWidth = ButtonWidth.fitContent,
      checked = false,
      classNames,
      configContextProps = {
        noDisabledContext: false,
        noSizeContext: false,
      },
      counter,
      disabled = false,
      disruptive = false,
      dropShadow = false,
      floatingButtonProps,
      nudgeProps: defaultNudgeProps,
      htmlType,
      iconProps,
      id,
      loading = false,
      onClick,
      onContextMenu,
      prefixIconProps,
      shape = ButtonShape.Pill,
      size = ButtonSize.Medium,
      split,
      splitButtonChecked = false,
      splitButtonProps,
      style,
      text,
      toggle,
      transparent = false,
      type, // TODO: Remove in Octuple v3.0.0 and use `variant` only.
      variant = ButtonVariant.Default,
      ...rest
    } = props;
    const largeScreenActive: boolean = useMatchMedia(Breakpoints.Large);
    const mediumScreenActive: boolean = useMatchMedia(Breakpoints.Medium);
    const smallScreenActive: boolean = useMatchMedia(Breakpoints.Small);
    const xSmallScreenActive: boolean = useMatchMedia(Breakpoints.XSmall);

    const internalRef: React.MutableRefObject<HTMLButtonElement> =
      useRef<HTMLButtonElement>(null);

    const mergedRef = useMergedRefs(internalRef, ref);

    const htmlDir: string = useCanvasDirection();

    const contextuallyDisabled: Disabled = useContext(DisabledContext);
    const mergedDisabled: boolean = configContextProps.noDisabledContext
      ? disabled
      : contextuallyDisabled || disabled;

    const contextuallySized: Size = useContext(SizeContext);
    const mergedSize = configContextProps.noSizeContext
      ? size
      : contextuallySized || size;

    const counterExists: boolean = !!counter;
    const iconExists: boolean = !!iconProps;
    const prefixIconExists: boolean = !!prefixIconProps;
    const textExists: boolean = !!text;

    const [nudgeProps, setNudgeProps] = useState<NudgeProps>(defaultNudgeProps);
    const innerNudgeRef: React.MutableRefObject<HTMLSpanElement> =
      useRef<HTMLSpanElement>(null);

    useEffect(() => {
      setNudgeProps(
        props.nudgeProps
          ? props.nudgeProps
          : {
              animation: NudgeAnimation.Background,
              delay: 2000,
              iterations: 1,
              enabled: false,
            }
      );
    }, [nudgeProps?.enabled]);

    useNudge(disruptive, nudgeProps, [internalRef, innerNudgeRef], styles);

    // TODO: Remove in Octuple v3.0.0 and use `variant` only.
    // For now, if `type` has a value use it, else use `variant`.
    const mergedVariant: ButtonVariant = !!type ? type : variant;

    const buttonSharedClassNames: string = mergeClasses([
      classNames,
      styles.button,
      { [styles.buttonDefault]: mergedVariant === ButtonVariant.Default },
      { [styles.buttonNeutral]: mergedVariant === ButtonVariant.Neutral },
      { [styles.buttonPrimary]: mergedVariant === ButtonVariant.Primary },
      { [styles.buttonSecondary]: mergedVariant === ButtonVariant.Secondary },
      { [styles.buttonSystemUi]: mergedVariant === ButtonVariant.SystemUI },
      {
        [styles.buttonDisruptive]:
          disruptive && mergedVariant === ButtonVariant.Default,
      },
      {
        [styles.buttonPrimaryDisruptive]:
          disruptive && mergedVariant === ButtonVariant.Primary,
      },
      {
        [styles.buttonSecondaryDisruptive]:
          disruptive && mergedVariant === ButtonVariant.Secondary,
      },
      {
        [styles.transparent]:
          transparent && mergedVariant === ButtonVariant.SystemUI,
      },
      {
        [styles.buttonSmall]:
          mergedSize === ButtonSize.Flex && largeScreenActive,
      },
      {
        [styles.buttonMedium]:
          mergedSize === ButtonSize.Flex && mediumScreenActive,
      },
      {
        [styles.buttonMedium]:
          mergedSize === ButtonSize.Flex && smallScreenActive,
      },
      {
        [styles.buttonLarge]:
          mergedSize === ButtonSize.Flex && xSmallScreenActive,
      },
      { [styles.buttonLarge]: mergedSize === ButtonSize.Large },
      { [styles.buttonMedium]: mergedSize === ButtonSize.Medium },
      { [styles.buttonSmall]: mergedSize === ButtonSize.Small },
      { [styles.pillShape]: shape === ButtonShape.Pill },
      {
        [styles.roundShape]:
          shape === ButtonShape.Round && !split && !textExists,
      },
      { [styles.dropShadow]: dropShadow },
      { [styles.disabled]: allowDisabledFocus || mergedDisabled },
      { [styles.floating]: floatingButtonProps?.enabled },
      { [styles.buttonRtl]: htmlDir === 'rtl' },
      {
        [styles.buttonConic]:
          !disruptive &&
          nudgeProps?.enabled &&
          nudgeProps?.animation === NudgeAnimation.Conic,
      },
    ]);

    const buttonClassNames: string = mergeClasses([
      buttonSharedClassNames,
      { [styles.buttonStretch]: buttonWidth === ButtonWidth.fill },
      { [styles.splitLeft]: split },
      { [styles.left]: alignText === ButtonTextAlign.Left },
      { [styles.right]: alignText === ButtonTextAlign.Right },
      {
        [styles.iconLeft]: iconExists && alignIcon === ButtonIconAlign.Left,
      },
      {
        [styles.iconRight]: iconExists && alignIcon === ButtonIconAlign.Right,
      },
      { [styles.loading]: loading },
    ]);

    const buttonTextClassNames: string = mergeClasses([
      {
        [styles.buttonTextSmall]:
          mergedSize === ButtonSize.Flex && largeScreenActive,
      },
      {
        [styles.buttonTextMedium]:
          mergedSize === ButtonSize.Flex && mediumScreenActive,
      },
      {
        [styles.buttonTextMedium]:
          mergedSize === ButtonSize.Flex && smallScreenActive,
      },
      {
        [styles.buttonTextLarge]:
          mergedSize === ButtonSize.Flex && xSmallScreenActive,
      },
      { [styles.buttonTextLarge]: mergedSize === ButtonSize.Large },
      { [styles.buttonTextMedium]: mergedSize === ButtonSize.Medium },
      { [styles.buttonTextSmall]: mergedSize === ButtonSize.Small },
    ]);

    const badgeClassNames: string = mergeClasses([
      styles.counter,
      buttonTextClassNames,
    ]);

    const getButtonIconSize = (): IconSize => {
      let iconSize: IconSize;
      if (mergedSize === ButtonSize.Flex && largeScreenActive) {
        iconSize = IconSize.Small;
      } else if (
        mergedSize === ButtonSize.Flex &&
        (mediumScreenActive || smallScreenActive)
      ) {
        iconSize = IconSize.Medium;
      } else if (mergedSize === ButtonSize.Flex && xSmallScreenActive) {
        iconSize = IconSize.Large;
      } else if (mergedSize === ButtonSize.Large) {
        iconSize = IconSize.Large;
      } else if (mergedSize === ButtonSize.Medium) {
        iconSize = IconSize.Medium;
      } else if (mergedSize === ButtonSize.Small) {
        iconSize = IconSize.Small;
      }
      return iconSize;
    };

    const getLoaderSize = (): LoaderSize => {
      let loaderSize: LoaderSize;
      if (size === ButtonSize.Flex && largeScreenActive) {
        loaderSize = LoaderSize.Small;
      } else if (
        size === ButtonSize.Flex &&
        (mediumScreenActive || smallScreenActive)
      ) {
        loaderSize = LoaderSize.Medium;
      } else if (size === ButtonSize.Flex && xSmallScreenActive) {
        loaderSize = LoaderSize.Large;
      } else if (size === ButtonSize.Large) {
        loaderSize = LoaderSize.Large;
      } else if (size === ButtonSize.Medium) {
        loaderSize = LoaderSize.Medium;
      } else if (size === ButtonSize.Small) {
        loaderSize = LoaderSize.Small;
      }
      return loaderSize;
    };

    const getButtonLoader = (): JSX.Element =>
      loading && (
        <Loader
          classNames={styles.loader}
          dotClassNames={styles.loaderDot}
          size={getLoaderSize()}
        />
      );

    const getButtonIcon = (): JSX.Element => (
      <Icon
        {...iconProps}
        classNames={mergeClasses([styles.icon, iconProps.classNames])}
        size={getButtonIconSize()}
      />
    );

    const getPrefixIcon = (): JSX.Element => (
      <Icon
        {...prefixIconProps}
        classNames={mergeClasses([
          styles.icon,
          styles.prefixIcon,
          prefixIconProps.classNames,
        ])}
        size={getButtonIconSize()}
      />
    );

    const getButtonContent = (
      buttonTextClassNames: string,
      text: string
    ): JSX.Element => (
      <span className={buttonTextClassNames}>
        {text ? text : 'Button'}
        {counterExists && <Badge classNames={badgeClassNames}>{counter}</Badge>}
      </span>
    );

    return (
      <>
        <button
          {...rest}
          ref={mergedRef}
          aria-checked={toggle ? !!checked : undefined}
          aria-disabled={mergedDisabled || loading}
          aria-label={ariaLabel}
          aria-pressed={toggle ? !!checked : undefined}
          defaultChecked={checked}
          disabled={(!allowDisabledFocus && mergedDisabled) || loading}
          className={buttonClassNames}
          id={id}
          onClick={!allowDisabledFocus ? onClick : null}
          style={style}
          type={htmlType}
        >
          <InnerNudge
            classNames={mergeClasses([
              styles.innerNudge,
              {
                [styles.conic]: nudgeProps?.animation === NudgeAnimation.Conic,
              },
            ])}
            disruptive={disruptive}
            id={id ? `${id}-nudge` : 'button-nudge'}
            nudgeProps={nudgeProps}
            ref={innerNudgeRef}
            style={style}
          />
          {iconExists && prefixIconExists && !textExists && (
            <span>
              {getButtonIcon()}
              {getPrefixIcon()}
            </span>
          )}
          {iconExists &&
            !textExists &&
            !prefixIconExists &&
            !counterExists &&
            getButtonIcon()}
          {counterExists && !textExists && !loading && !iconExists && (
            <Badge classNames={badgeClassNames}>{counter}</Badge>
          )}
          {iconExists && counterExists && !textExists && !loading && (
            <span>
              {getButtonIcon()}
              <Badge classNames={badgeClassNames}>{counter}</Badge>
              {prefixIconExists && getPrefixIcon()}
            </span>
          )}
          {iconExists && textExists && (
            <span>
              {getButtonIcon()}
              {getButtonContent(buttonTextClassNames, text)}
              {prefixIconExists && getPrefixIcon()}
            </span>
          )}
          {!iconExists && getButtonContent(buttonTextClassNames, text)}
          {getButtonLoader()}
        </button>
        {split && (
          <SplitButton
            {...splitButtonProps}
            classNames={
              buttonSharedClassNames + ' ' + splitButtonProps?.classNames
            }
            checked={splitButtonChecked}
            disruptive={disruptive}
            dropShadow={dropShadow}
            nudgeProps={nudgeProps}
            onClick={
              !splitButtonProps?.allowDisabledFocus ? onContextMenu : null
            }
            shape={shape}
            size={mergedSize}
            split={split}
            variant={mergedVariant}
          />
        )}
      </>
    );
  }
);
