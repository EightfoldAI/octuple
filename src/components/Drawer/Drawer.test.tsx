import React, { useState } from 'react';
import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import MatchMediaMock from 'jest-matchmedia-mock';
import { Drawer, DrawerVariant } from './';
import { Button, ButtonVariant, ButtonWidth } from '../Button';
import { IconName } from '../Icon';
import { fireEvent, render } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import '@testing-library/jest-dom';

Enzyme.configure({ adapter: new Adapter() });

let matchMedia: any;

type DrawerTestProps = {
  onClose?: () => void;
  onDraggerClick?: () => void;
  onDraggerKeyDown?: () => void;
  variant?: DrawerVariant;
};

const DrawerTest: React.FC<DrawerTestProps> = ({
  onClose = () => {},
  onDraggerClick = () => {},
  onDraggerKeyDown = () => {},
  variant = DrawerVariant.Default,
}) => {
  const [visible, setVisible] = useState<boolean>(false);
  const _onClose = (_event: React.MouseEvent<HTMLButtonElement>) => {
    if (!visible) {
      return;
    }
    setVisible(false);
  };
  const _onOpen = (_event: React.MouseEvent<HTMLButtonElement>) => {
    if (visible) {
      return;
    }
    setVisible(true);
  };
  return (
    <>
      <Button
        onClick={visible ? (e) => _onClose(e) : (e) => _onOpen(e)}
        text={visible ? 'Close drawer' : 'Open drawer'}
        variant={ButtonVariant.Primary}
      />
      <Drawer
        children={<div>Test drawer</div>}
        data-testid="testDrawer"
        footer={
          <div>
            <Button
              buttonWidth={ButtonWidth.fill}
              onClick={(e) => _onClose(e)}
              text="Close"
              variant={ButtonVariant.Primary}
            />
          </div>
        }
        variant={variant}
        visible={visible}
        overlay={visible}
        maskClosable={visible}
        onVisibleChange={(v) => setVisible(v)}
        onClose={onClose}
        onDraggerClick={onDraggerClick}
        onDraggerKeyDown={onDraggerKeyDown}
      />
    </>
  );
};

describe('Drawer', () => {
  beforeAll(() => {
    matchMedia = new MatchMediaMock();
  });

  afterEach(() => {
    matchMedia.clear();
  });

  test('Drawer renders', () => {
    const { container } = render(
      <Drawer children={<div>Test drawer</div>} data-testid="testDrawer" />
    );
    expect(() => container).not.toThrowError();
    expect(container).toMatchSnapshot();
  });

  test('Renders with default props', () => {
    const { getByTestId } = render(
      <Drawer children={<div>Test drawer</div>} data-testid="testDrawer" />
    );
    expect(getByTestId('testDrawer')).toBeTruthy();
  });

  test('Handles open and close', () => {
    const { getByText } = render(<DrawerTest />);
    fireEvent.click(getByText('Open drawer'));
    expect(getByText('Close drawer')).toBeTruthy();
    fireEvent.click(getByText('Close drawer'));
    expect(getByText('Open drawer')).toBeTruthy();
  });

  test('Handles DrawerVariant.Default', () => {
    const { getByTestId } = render(
      <Drawer
        children={<div>Test drawer</div>}
        data-testid="testDrawer"
        variant={DrawerVariant.Default}
      />
    );
    expect(
      getByTestId('testDrawer').classList.contains('modeless')
    ).toBeFalsy();
  });

  test('Handles DrawerVariant.Default modeless', () => {
    const { getByTestId } = render(
      <Drawer
        children={<div>Test drawer</div>}
        data-testid="testDrawer"
        variant={DrawerVariant.Default}
        maskClosable={false}
        overlay={false}
      />
    );
    expect(
      getByTestId('testDrawer').classList.contains('modeless')
    ).toBeTruthy();
  });

  test('Handles DrawerVariant.Hint', () => {
    const { getByTestId } = render(
      <Drawer
        children={<div>Test drawer</div>}
        data-testid="testDrawer"
        variant={DrawerVariant.Hint}
      />
    );
    expect(
      getByTestId('testDrawer').classList.contains('modeless')
    ).toBeFalsy();
  });

  test('Handles DrawerVariant.Hint modeless', () => {
    const { getByTestId } = render(
      <Drawer
        children={<div>Test drawer</div>}
        data-testid="testDrawer"
        variant={DrawerVariant.Hint}
        maskClosable={false}
        overlay={false}
      />
    );
    expect(
      getByTestId('testDrawer').classList.contains('modeless')
    ).toBeTruthy();
  });

  test('Drawer visibility', () => {
    const { getByTestId, rerender } = render(
      <Drawer
        children={<div>Test drawer</div>}
        data-testid="testDrawer"
        variant={DrawerVariant.Default}
        visible
      />
    );
    expect(getByTestId('testDrawer').classList.contains('visible')).toBe(true);
    rerender(
      <Drawer
        children={<div>Test drawer</div>}
        data-testid="testDrawer"
        variant={DrawerVariant.Default}
        visible={false}
      />
    );
    expect(getByTestId('testDrawer').classList.contains('visible')).toBe(false);
  });

  test('Drawer content', () => {
    const { getByTestId } = render(
      <Drawer
        children={<div>Test drawer content</div>}
        title={<h1>Test drawer header</h1>}
        data-testid="testDrawer"
        variant={DrawerVariant.Default}
        visible
      />
    );
    expect(
      getByTestId('testDrawer').getElementsByClassName('body')[0].innerHTML
    ).toBe('<div>Test drawer content</div>');
    expect(
      getByTestId('testDrawer').getElementsByClassName('header-title')[0]
        .innerHTML
    ).toBe('<h1>Test drawer header</h1>');
  });

  test('Drawer onClose action', () => {
    const onEvent = jest.fn();
    const { container, getByTestId } = render(
      <Drawer
        children={<div>Test drawer content</div>}
        title={<h1>Test drawer header</h1>}
        data-testid="testDrawer"
        onClose={onEvent}
        variant={DrawerVariant.Hint}
        visible
      />
    );
    expect(
      getByTestId('testDrawer').getElementsByClassName('body')[0].innerHTML
    ).toBe('<div>Test drawer content</div>');
    act(() => {
      fireEvent.click(
        getByTestId('testDrawer').getElementsByClassName('dragger')[0]
      );
    });
    expect(onEvent).toHaveBeenCalled();
  });

  test('Drawer onDraggerClick action', () => {
    const onEvent = jest.fn();
    const { getByTestId } = render(
      <Drawer
        children={<div>Test drawer content</div>}
        title={<h1>Test drawer header</h1>}
        data-testid="testDrawer"
        onDraggerClick={onEvent}
        variant={DrawerVariant.Default}
        visible
      />
    );
    expect(
      getByTestId('testDrawer').getElementsByClassName('body')[0].innerHTML
    ).toBe('<div>Test drawer content</div>');
    act(() => {
      fireEvent.click(
        getByTestId('testDrawer').getElementsByClassName('dragger')[0]
      );
    });
    expect(onEvent).toHaveBeenCalled();
  });

  test('Drawer onDraggerKeyDown action', () => {
    const onEvent = jest.fn();
    const { getByTestId } = render(
      <Drawer
        children={<div>Test drawer content</div>}
        title={<h1>Test drawer header</h1>}
        data-testid="testDrawer"
        onDraggerKeyDown={onEvent}
        variant={DrawerVariant.Default}
        visible
      />
    );
    expect(
      getByTestId('testDrawer').getElementsByClassName('body')[0].innerHTML
    ).toBe('<div>Test drawer content</div>');
    act(() => {
      fireEvent.keyDown(
        getByTestId('testDrawer').getElementsByClassName('dragger')[0],
        { key: 'Enter', code: 'Enter' }
      );
    });
    expect(onEvent).toHaveBeenCalled();
  });

  test('Drawer header actions exist', () => {
    const { getByTestId } = render(
      <Drawer
        children={<div>Test drawer</div>}
        data-testid="testDrawer"
        actionButtonOneProps={{
          classNames: 'header-action-button-1',
          iconProps: { path: IconName.mdiCogOutline },
        }}
        actionButtonTwoProps={{
          classNames: 'header-action-button-2',
          iconProps: { path: IconName.mdiHistory },
        }}
        actionButtonThreeProps={{
          classNames: 'header-action-button-3',
          iconProps: { path: IconName.mdiDatabaseArrowDownOutline },
        }}
        variant={DrawerVariant.Hint}
        visible
      />
    );
    expect(
      getByTestId('testDrawer').getElementsByClassName(
        'header-action-button-1'
      )[0]
    ).toBeTruthy();
    expect(
      getByTestId('testDrawer').getElementsByClassName(
        'header-action-button-2'
      )[0]
    ).toBeTruthy();
    expect(
      getByTestId('testDrawer').getElementsByClassName(
        'header-action-button-3'
      )[0]
    ).toBeTruthy();
  });

  test('Drawer no body padding', () => {
    const { container, getByTestId } = render(
      <Drawer
        children={<div>Test drawer</div>}
        data-testid="testDrawer"
        bodyPadding={false}
        variant={DrawerVariant.Default}
        visible
      />
    );
    expect(
      getByTestId('testDrawer').getElementsByClassName('no-body-padding')[0]
    ).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  test('Drawer no header padding', () => {
    const { container, getByTestId } = render(
      <Drawer
        children={<div>Test drawer content</div>}
        title={<h1>Test drawer header</h1>}
        data-testid="testDrawer"
        headerPadding={false}
        variant={DrawerVariant.Default}
        visible
      />
    );
    expect(
      getByTestId('testDrawer').getElementsByClassName('no-header-padding')[0]
    ).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  test('Drawer overlay is hidden', () => {
    const { getByTestId } = render(
      <Drawer
        children={<div>Test drawer content</div>}
        title={<h1>Test drawer header</h1>}
        data-testid="testDrawer"
        overlay={false}
        variant={DrawerVariant.Default}
        visible
      />
    );
    expect(
      getByTestId('testDrawer').classList.contains('modeless')
    ).toBeTruthy();
  });

  test('Drawer footer', () => {
    const { getByTestId } = render(<DrawerTest />);
    expect(
      getByTestId('testDrawer').getElementsByClassName('footer')[0]
    ).toBeTruthy();
    expect(
      getByTestId('testDrawer').getElementsByClassName('button-primary')[0]
    ).toBeTruthy();
  });

  test('Should render content when renderContentAlways is true', () => {
    const { getByText } = render(
      <Drawer
        children={<div>Content is always rendered</div>}
        data-testid="testDrawer"
        renderContentAlways
        variant={DrawerVariant.Default}
      />
    );
    expect(getByText('Content is always rendered')).toBeTruthy();
  });

  test('Should not render content when renderContentAlways is false', () => {
    const { queryByText } = render(
      <Drawer
        children={<div>Content is not always rendered</div>}
        data-testid="testDrawer"
        renderContentAlways={false}
        variant={DrawerVariant.Default}
      />
    );
    expect(queryByText('Content is not always rendered')).toBeNull();
  });
});
