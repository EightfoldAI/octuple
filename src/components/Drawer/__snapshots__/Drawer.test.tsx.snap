// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer Drawer no body padding 1`] = `<div />`;

exports[`Drawer Drawer no header padding 1`] = `<div />`;

exports[`Drawer Drawer renders 1`] = `<div />`;
